### PR TITLE
Allow providing of actual datum for reference inputs

### PR DESCRIPTION
**Pull request by carbolymer**
# Changelog

```yaml
- description: |
    Allow providing of actual datum for reference inputs in `TxInsReference`.
# uncomment types applicable to the change:
  type:
   - feature        # introduces a new feature
   - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - refactoring    # QoL changes
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

This PR adds a method of adding actual datum for the datum hashes present in reference inputs.

Tested in: 
- https://github.com/IntersectMBO/cardano-node/pull/6174

Integrated in CLI:
- https://github.com/IntersectMBO/cardano-cli/pull/1169

Closes https://github.com/IntersectMBO/cardano-api/issues/803

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Internal/Tx/Body.hs [hunk 1]
```diff
@@ -1742,11 +1761,11 @@ fromLedgerTxInsCollateral sbe body =
     sbe
 
 fromLedgerTxInsReference
-  :: ShelleyBasedEra era -> Ledger.TxBody (ShelleyLedgerEra era) -> TxInsReference era
+  :: ShelleyBasedEra era -> Ledger.TxBody (ShelleyLedgerEra era) -> TxInsReference ViewTx era
 fromLedgerTxInsReference sbe txBody =
   caseShelleyToAlonzoOrBabbageEraOnwards
     (const TxInsReferenceNone)
-    (\w -> TxInsReference w $ map fromShelleyTxIn . toList $ txBody ^. L.referenceInputsTxBodyL)
+    (\w -> TxInsReference w (map fromShelleyTxIn . toList $ txBody ^. L.referenceInputsTxBodyL) ViewTx)
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Internal/Tx/Body.hs:1768]: Ah yes, there is now way to figure out which datum hashes are relevant.

File: cardano-api/src/Cardano/Api/Internal/Tx/Body.hs [hunk 2]
```diff
@@ -3053,17 +3081,30 @@ collectTxBodyScriptWitnessRequirements
 
 getSupplementalDatums
   :: AlonzoEraOnwards era
+  -> TxInsReference BuildTx era
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Internal/Tx/Body.hs:3084]: These technically wouldn't be supplemental datums as those correspond to tx outputs.

**Comment by carbolymer** [comment]
@mmontin This should implement #803 . What do you think about the design here?

File: cardano-api/src/Cardano/Api/Internal/Tx/Body.hs [hunk 3]
```diff
@@ -570,16 +572,20 @@ deriving instance Eq (TxTotalCollateral era)
...
-data TxInsReference era where
-  TxInsReferenceNone :: TxInsReference era
+data TxInsReference build era where
+  TxInsReferenceNone :: TxInsReference build era
   TxInsReference
     :: BabbageEraOnwards era
     -> [TxIn]
-    -> TxInsReference era
+    -- ^ A list of reference inputs
+    -> BuildTxWith build (Set HashableScriptData)
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Internal/Tx/Body.hs]: I like this: `type TxInsReferenceActualDatums build = BuildTxWith build (Set HashableScriptData)` 

It's more concise and we have the notion of empty via the `Set` type. Go ahead with the change and I'll have a look after.

**Review by Jimbo4350 — APPROVED** [review]
LGTM :+1: . Get input from the ledger team regarding naming and we can finalize that aspect.

File: cardano-api/src/Cardano/Api/Internal/Tx/Output.hs [hunk 4]
```diff
@@ -14,30 +14,43 @@
 {-# LANGUAGE TypeOperators #-}
 
 module Cardano.Api.Internal.Tx.Output
```
Comment (mmontin) [inline cardano-api/src/Cardano/Api/Internal/Tx/Output.hs:16]: Changes in this file seem to be unrelated to the global PR. I don't mind it too much personally but I don't know what the practice are for cardano-api. I imagine you do ;)

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs [hunk 5]
```diff
@@ -294,7 +294,7 @@ prop_make_transaction_body_autobalance_return_correct_fee_for_multi_asset = H.pr
         Nothing
 
   scriptWitReqsWithAsset <-
-    H.evalEither $ collectTxBodyScriptWitnessRequirements aeo balancedContentWithTxoutAsset
+    H.evalEither $ collectTxBodyScriptWitnessRequirements aeo utxos balancedContentWithTxoutAsset
```
Comment (mmontin) [inline cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs:297]: I guess this could be `empty` instead of `utxos` as it is a singleton that was built for other purposes. I think this would be more understandable and not make it look as if it is relevant here.

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs [hunk 6]
```diff
@@ -324,7 +324,7 @@ prop_make_transaction_body_autobalance_return_correct_fee_for_multi_asset = H.pr
         Nothing
 
   scriptWitReqsBalanced <-
-    H.evalEither $ collectTxBodyScriptWitnessRequirements aeo balancedContent
+    H.evalEither $ collectTxBodyScriptWitnessRequirements aeo utxos balancedContent
```
Comment (mmontin) [inline cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs:327]: same as above

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs [hunk 7]
```diff
@@ -453,7 +453,7 @@ prop_make_transaction_body_autobalance_multi_asset_collateral = H.propertyOnce $
...
```
Comment (mmontin) [inline cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs:456]: same

**Review by mmontin — COMMENTED** [review]
Thanks a lot for implementing this ! I believe this feature, while somewhat a corner case, is very important to have.

In terms of the actual implementation, I have two general comments / remarks on top of the small comments I left here and there

1. I wonder why these additional datums are given as a set directly inside `TxInsReference` instead of input by input. This breaks the link as to which input is supposed to bear which datum hash. In other words, why not couple each `TxIn` with a `Maybe HashableScriptData` in a structure (let's call it `RefInputWithDatum`) and have `TxInsReference` take a list of those? 

2. About the need of providing a `UTxO` here and there. I understand the need for this structure. It allows to ensure that the provided datums do indeed correspond to the resolved variants of hashed datums in reference inputs. However, this will be caught later on in the pipeline anyway as the ledger with throw an error when they do not. So is this necessary to ensure this here? Admittedly, I don't have a full picture as to which properties need to be satisfied by `createTransactionBody` and are expected to hold by the caller. However, it makes the implementation more involved, and also forces the caller to build this UTxO and understand that it corresponds to the corner case of providing resolved datums to reference inputs. Additionally, since the current implementation seems to just ignore datums that do not correspond to hashed datums in ref inputs, there is no clear way to bypass this requirement for testing purposes. And I believe this is dangerous that this process is silent. This makes me wonder if there are other places in the code base where silent discards are performed.

File: cardano-api/src/Cardano/Api/Internal/Tx/Output.hs (hunk 4 above)
Comment (carbolymer) [reply cardano-api/src/Cardano/Api/Internal/Tx/Output.hs:16]: Ah yes, those are unrelated. I've addressed one remark from other PR. I'll make a separate commit with that.

**Comment by carbolymer** [comment]
@mmontin Good remarks thanks!

ad 2. I agree. Passing UTXO around mostly gets in the way, without any real benefit here. I think we could skip it, ditch the filtering and do the validation at transaction submission step (by ledger).

ad 1. Assuming we remove the necessity of passing UTXO: tying `TxIns` with datum values would only make sense if we have UTXOs of reference inputs in the scope, and we would validate that respective datum is at the right input. I agree that there's a link there, but I'm not sure there's a value in that information: this link gets "lost" when the transaction gets submitted to the ledger. So the constructor `TxInsReference` would potentially look similar to this:
```haskell
 TxInsReference
    :: BabbageEraOnwards era
    -> [( TxIn
        , BuildTx build (Maybe HashableScriptData)
        )
       ]
    -> TxInsReference build era
```
which would complicate handling of the type a bit: the values would have to be `zip`ped and `unzip`ed when constructing and deconstructing the type.

**Comment by mmontin** [comment]
About location of datums in the body. Your point makes sense, it would be unwise to complicate the type for no valid reason. And to be honest, the current version where the datum are grouped into a set would be suitable enough for my use case and overall satisfactory, in my opinion. However, I see 2 reasons why the other option also makes sense, based on my current understanding of cardano-api and my current biases as to how I use it in `cooked-validators`:

1. **Consistency**: Currently, the second use case for supplemental datums (when paying outputs with hashed datums) is handled locally, within the `TxOut` itself. Technically, `TxOutSupplementalDatum` is the same as `TxOutDatumHash` with the exception that the full datum will be given to the transaction. This means that `TxOutSupplementalDatum` could technically be removed and, instead, the `txOuts` field of the `TxBodyContent` could be typed:
```
TxOuts
   :: BabbageEraOnwards era -- not sure about Babbage, but not so relevant for the example
   -> [TxOut ctx era]
   -> Set (BuildTx build HashableScriptData)
   -> TxOuts build era
```
I am definitely _not_ advocating for such a change, just pointing out that the choice was made to have these supplemental datums be handled within each `TxOut` instead of grouped outside. Admittedly, this makes a lot more sense here, because the `TxOut`s naturally contain a datum, as opposed to `TxIn` which are merely ids. That being said, handling supplemental datums similarly for the second possible use case would make sense and likely be more familiar to future users relying on this new feature. 

2. **Locality**: This one is more related to my personal bias, but I like when the information is as close as possible to where it belongs. Granted, this will be lost later on in the pipeline, but I see cardano-api transaction bodies as an opportunity to layout pieces of information in a more local way as the transaction produced by the ledger and given to the scripts. Having these supplemental datums defined locally would _feel_ just right (for me at least) and this is the place where I would naturally search for this information. That being said, having the datum a bit higher up does not hurt and they could also be found quite easily there.

I feel like this second point is not quite as strong as I thought when I decided to write it, because it's mostly just based on a feeling, and I leave it to you to take it into account or not. Regardless, I hope these points are of help in making the final decision.

**Comment by Jimbo4350** [comment]
> 1. I wonder why these additional datums are given as a set directly inside `TxInsReference` instead of input by input. This breaks the link as to which input is supposed to bear which datum hash. In other words, why not couple each `TxIn` with a `Maybe HashableScriptData` in a structure (let's call it `RefInputWithDatum`) and have `TxInsReference` take a list of those?

What would be the purpose? You would lose the connection as soon as you construct a valid transaction.

>  However, this will be caught later on in the pipeline anyway as the ledger with throw an error when they do not.

I agree with this regarding removing the `UTxO era` parameter. cc: @carbolymer

**Comment by Jimbo4350** [comment]
>  just pointing out that the choice was made to have these supplemental datums be handled within each TxOut instead of grouped outside.

I see your point but the `TxOut` type was designed this way so that consumers of cardano-api could easily differentiate between the different datums that exist at outputs.

I wouldn't include the "resolved datums'  in the `TxOut` definition because it is involved with the creation of datums/datum hashes (or not) at outputs. The "resolved datums" feature is about including already existing datums in the datum map.

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs (hunk 5 above)
Comment (carbolymer) [reply cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs:297]: Removed `utxos` parameter.

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs (hunk 7 above)
Comment (carbolymer) [reply cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs:456]: Removed `utxos` parameter.

File: cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs (hunk 6 above)
Comment (carbolymer) [reply cardano-api/test/cardano-api-test/Test/Cardano/Api/Transaction/Autobalance.hs:327]: Removed `utxos` argument.

File: cardano-api/src/Cardano/Api/Internal/Tx/Body.hs [hunk 8]
```diff
@@ -1075,25 +1095,35 @@ addTxInCollateral
...
+      TxInsReference era xs bScriptData' -> TxInsReference era (xs <> txInsReference) ((<> scriptData) <$> bScriptData')
 
 addTxInReference
-  :: IsBabbageBasedEra era => TxIn -> TxBodyContent build era -> TxBodyContent build era
-addTxInReference txInReference = addTxInsReference [txInReference]
+  :: Applicative (BuildTxWith build)
+  => IsBabbageBasedEra era
+  => TxIn
+  -> TxBodyContent build era
+  -> TxBodyContent build era
```
[the remark this answers is not in the document]
Comment (carbolymer) [reply cardano-api/src/Cardano/Api/Internal/Tx/Body.hs:1125]: Good point, added, thanks.
